### PR TITLE
Basic UI Indicate Ambiguous/Impossible Commands

### PR DESCRIPTION
**Pull request by jcolemang**
Currently Just an alert. I could work on adding more of an actual UI (which would be particularly useful for when we want to let the user give feedback) but I figured that should go into a new PR.

#38

### CODE REVIEW

File: server/neural_network_interface.py [hunk 1]
```diff
@@ -32,10 +32,11 @@ def generate_move(self, sid, gesture_data, message_data):
                 movement_ct)
 
 class NeuralNetworkBlocksworldModel(BlocksworldModel):
-    def __init__(self, h5_paths):
+    def __init__(self, h5_paths, ambiguity_threshold=0.3):
         (self.flip_model, self.colors_model, self.letters_model) = runner.load_models(h5_paths)
         self.tokenizer = core.build_tokenizer(core.load_vocabulary())
-        BlocksworldModel.__init__(self)
+        super().__init__()
```
Comment (Lewis-Kelley) [inline server/neural_network_interface.py]: This line is causing a build error (the previous way is the way `pylint` accepts). Also, isn't this traditionally at the top or the bottom?

File: 2D Simulation/game.js [hunk 2]
```diff
@@ -160,6 +140,27 @@ socket.on('update_flip_block', function (block_id) {
     flipBlock(block_id, null, currentConfig);
 });
 
+socket.on('indicate_impossible_move', function(move) {
```
Comment (Lewis-Kelley) [inline 2D Simulation/game.js]: If you're adding new Socket.io signals, you need to update the sequence diagrams in the docs.

File: server/neural_network_interface.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 from rhit_xprize_neural_network import model_runner as runner
 from rhit_xprize_neural_network import trainer_core as core
 
-class BlocksworldModel:
+class BlocksworldModel(object):
```
Comment (Lewis-Kelley) [inline server/neural_network_interface.py:9]: Does this change anything?

File: server/neural_network_interface.py [hunk 4]
```diff
@@ -46,14 +47,18 @@ def generate_move(self, sid, gesture_data, message_data):
 
         (flip, color, letter) = self._run_models(message)
```
Comment (Lewis-Kelley) [inline server/neural_network_interface.py]: I realize this is technically a "my bad," but can we rename these to `flip_tuple` or `flip_tup` or something like that to emphasize they hold both a flip/color/letter value _and_ a "probability"?

File: server/neural_network_interface.py [hunk 5]
```diff
@@ -77,6 +82,21 @@ def _build_flip(self, sid, block_id):
             'move_number': self._movement_count_dict[sid]
         }
 
+    def _build_ambiguous(self, letter, color, block_candidates):
+        return {
+            'type': 'ambiguous',
+            'predicted_letter': letter if letter != 'None' else False,
+            'predicted_color': color if color != 'None' else False,
```
Comment (Lewis-Kelley) [inline server/neural_network_interface.py:89]: What happens if we pass `None` instead of `False`? It makes more sense logically, but I'm not sure how that would be received on the JavaScript side.

File: server/neural_network_interface.py [hunk 6]
```diff
@@ -77,6 +82,21 @@ def _build_flip(self, sid, block_id):
...
+            'candidates': block_candidates
+        }
+
+    def _build_impossible(self, letter, color):
+        return {
+            'type': 'impossible',
+            'predicted_letter': letter if letter != 'None' else False,
```
Comment (Lewis-Kelley) [inline server/neural_network_interface.py:96]: Ditto on the `None` vs. `False` thing.

File: server/tests/neural_network_interface_test.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    letter = 'x'
+    game_state = [
+        construct_block('#block1', 'red', 'a', 'blue', 'b'),
+        construct_block('#block2', 'purple', 'a', 'blue', 'c')
+    ]
+
+    result = nni._find_block(game_state, color, letter)
+    assert result == []
+    
+def test_find_block_incorrect_color_no_letters():
```
Comment (Lewis-Kelley) [inline server/tests/neural_network_interface_test.py]: This test has the same name as the previous...so I think it would shadow it.

File: server/tests/neural_network_interface_test.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    color = 'green'
+    letter = 'z'
+    game_state = [
+        construct_block('#block1', 'red', 'a', 'green', 'b'),
+        construct_block('#block2', 'beije', 'a', 'yellow', 'c')
```
Comment (Lewis-Kelley) [inline server/tests/neural_network_interface_test.py]: FYI: it's "beige."

File: server/tests/neural_network_interface_test.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+def test_find_block_color_one_correct_letter():
+    color = 'none'
+    letter = 'b'
+    game_state = [
+        construct_block('block1', 'red', 'a', 'blue', 'b'),
+        construct_block('block2', 'purple', 'b', 'blue', 'c')
+    ]
+
+    result = nni._find_block(game_state, color, letter)
+    assert result == ['block2']
```
Comment (Lewis-Kelley) [inline server/tests/neural_network_interface_test.py]: We should probably also have a test for what happens with (`none`, `none`) and where there are multiple valid blocks.

File: 2D Simulation/game.js [hunk 10]
```diff
@@ -143,10 +123,10 @@ socket.on('enable_blocks_for_player_2', function(data) {
 
 var z = 1;
 socket.on('update_position', function (data) {
-    var left = Math.max(data.left, 0);
-    var top = Math.max(data.top, 0);
+    let left = Math.max(data.left, 0);
+    let top = Math.max(data.top, 0);
 
-    var block = data.block_id;
+    let block = data.block_id;
```
Comment (Lewis-Kelley) [inline 2D Simulation/game.js]: ...and JavaScript said, "`let` there be blocks!" and there were.

**Review by Lewis-Kelley — CHANGES_REQUESTED** [review]
Yay tests! Looks good in general. I've got a some small suggestions, but the one I want to emphasize is that the build is broken.

For whatever reason, `pylint` does not like the `super().__init__()` calls. If you can find a way around that, great. Otherwise, we're kind of stuck with what I had before.

File: server/neural_network_interface.py (hunk 3 above)
Comment (jcolemang) [reply server/neural_network_interface.py:9]: Fixed a linter error

**Comment by jcolemang** [comment]
I haven't looked through all the feedback yet, but as a side note I am still getting back to the linter errors.

File: server/neural_network_interface.py (hunk 5 above)
Comment (jcolemang) [reply server/neural_network_interface.py:89]: Nothing exciting happens, I would just have to check for `'None'` instead of false, but I disagree that it makes more sense logically. Passing a falsey value to indicate that there is no color seems more natural to me.

File: server/neural_network_interface.py (hunk 6 above)
Comment (jcolemang) [reply server/neural_network_interface.py:96]: ack

File: server/tests/neural_network_interface_test.py (hunk 7 above)
Comment (jcolemang) [reply server/tests/neural_network_interface_test.py]: It also basically has the same body. Not sure what I was doing there.

File: server/neural_network_interface.py (hunk 4 above)
Comment (jcolemang) [reply server/neural_network_interface.py]: ack

File: 2D Simulation/game.js (hunk 2 above)
Comment (jcolemang) [reply 2D Simulation/game.js]: If I am reading them correctly they are already out of date, I think that the AI mode should be added in a separate PR.

File: server/tests/neural_network_interface_test.py (hunk 8 above)
Comment (jcolemang) [reply server/tests/neural_network_interface_test.py]: ack

File: server/tests/neural_network_interface_test.py (hunk 9 above)
Comment (jcolemang) [reply server/tests/neural_network_interface_test.py]: ack

**Comment by jcolemang** [comment]
Note after addressing comments, something in the linter is incorrect. I am adding an issue to address it. For this PR I just took out the linter.

File: 2D Simulation/game.js (hunk 2 above)
Comment (Lewis-Kelley) [reply 2D Simulation/game.js]: There is the AI mode sequence diagram in #42. I'll quickly pull that into its own PR and get that pushed up.

File: server/neural_network_interface.py (hunk 5 above)
Comment (Lewis-Kelley) [reply server/neural_network_interface.py:89]: I was thinking the keyword `None` rather than the String value `'None'`, unless JavaScript sees it on the other end as the string `'None'`, in which case yes that doesn't really improve anything. I'd hope it would come out the other end as `undefined`.

**Review by Lewis-Kelley — APPROVED** [review]

File: server/neural_network_interface.py (hunk 5 above)
Comment (jcolemang) [reply server/neural_network_interface.py:89]: My guess is that it would be `'None'`, but I wouldn't really like `undefined` either because I think that it is just meant for undefined variables.

File: .travis.yml [hunk 11]
```diff
@@ -7,6 +7,5 @@ install:
   - sudo pip install -r requirements.txt
 
 script:
-  - sudo python -m pytest . --ignore=rhit-xprize-neural-network
   - cd server
-  - pylint -E *.py
```
Comment (kgroble) [inline .travis.yml:12]: Did you ever figure this out? I'm fine leaving it out for now, but it'd be nice to know what's going on.

**Review by kgroble — APPROVED** [review]

File: .travis.yml (hunk 11 above)
Comment (Lewis-Kelley) [reply .travis.yml:12]: This is brought up in #44.

Comment (jcolemang) [reply .travis.yml:12]: @kgroble I didn't ever figure it out. I think that travis is just trying very hard to make my life unpleasant.

**Review by kgroble — APPROVED** [review]

**Comment by kgroble** [comment]
Oops. Guess I doubly approve of this.
